// @zetachain/toolkit/dist/tasks/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyTask = exports.sendZRC20Task = exports.sendZETATask = exports.sendBTCTask = exports.poolsTask = exports.omnichainTask = exports.messagingTask = exports.feesTask = exports.faucetTask = exports.cctxTask = exports.balancesTask = exports.accountTask = void 0;
var account_1 = require("./account");
Object.defineProperty(exports, "accountTask", { enumerable: true, get: function () { return account_1.accountTask; } });
var balances_1 = require("./balances");
Object.defineProperty(exports, "balancesTask", { enumerable: true, get: function () { return balances_1.balancesTask; } });
var cctx_1 = require("./cctx");
Object.defineProperty(exports, "cctxTask", { enumerable: true, get: function () { return cctx_1.cctxTask; } });
var faucet_1 = require("./faucet");
Object.defineProperty(exports, "faucetTask", { enumerable: true, get: function () { return faucet_1.faucetTask; } });
var fees_1 = require("./fees");
Object.defineProperty(exports, "feesTask", { enumerable: true, get: function () { return fees_1.feesTask; } });
var messaging_1 = require("./messaging");
Object.defineProperty(exports, "messagingTask", { enumerable: true, get: function () { return messaging_1.messagingTask; } });
var omnichain_1 = require("./omnichain");
Object.defineProperty(exports, "omnichainTask", { enumerable: true, get: function () { return omnichain_1.omnichainTask; } });
var pools_1 = require("./pools");
Object.defineProperty(exports, "poolsTask", { enumerable: true, get: function () { return pools_1.poolsTask; } });
var sendBTC_1 = require("./sendBTC");
Object.defineProperty(exports, "sendBTCTask", { enumerable: true, get: function () { return sendBTC_1.sendBTCTask; } });
var sendZETA_1 = require("./sendZETA");
Object.defineProperty(exports, "sendZETATask", { enumerable: true, get: function () { return sendZETA_1.sendZETATask; } });
var sendZRC20_1 = require("./sendZRC20");
Object.defineProperty(exports, "sendZRC20Task", { enumerable: true, get: function () { return sendZRC20_1.sendZRC20Task; } });
var verify_1 = require("./verify");
Object.defineProperty(exports, "verifyTask", { enumerable: true, get: function () { return verify_1.verifyTask; } });
