// @inquirer/core/dist/cjs/lib/screen-manager.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cli_width_1 = __importDefault(require("cli-width"));
const strip_ansi_1 = __importDefault(require("strip-ansi"));
const string_width_1 = __importDefault(require("string-width"));
const ansi_escapes_1 = __importDefault(require("ansi-escapes"));
const util = __importStar(require('./readline.js'));
const utils_mjs_1 = require('./utils.js');
const height = (content) => content.split('\n').length;
const lastLine = (content) => { var _a; return (_a = content.split('\n').pop()) !== null && _a !== void 0 ? _a : ''; };
class ScreenManager {
    constructor(rl) {
        Object.defineProperty(this, "rl", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: rl
        });
        // These variables are keeping information to allow correct prompt re-rendering
        Object.defineProperty(this, "height", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        Object.defineProperty(this, "extraLinesUnderPrompt", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 0
        });
        this.rl = rl;
    }
    render(content, bottomContent = '') {
        this.clean();
        this.rl.output.unmute();
        /**
         * Write message to screen and setPrompt to control backspace
         */
        const promptLine = lastLine(content);
        const rawPromptLine = (0, strip_ansi_1.default)(promptLine);
        // Remove the rl.line from our prompt. We can't rely on the content of
        // rl.line (mainly because of the password prompt), so just rely on it's
        // length.
        let prompt = rawPromptLine;
        if (this.rl.line.length) {
            prompt = prompt.slice(0, -this.rl.line.length);
        }
        this.rl.setPrompt(prompt);
        // SetPrompt will change cursor position, now we can get correct value
        const cursorPos = this.rl._getCursorPos();
        const width = (0, cli_width_1.default)({ defaultWidth: 80, output: this.rl.output });
        content = (0, utils_mjs_1.breakLines)(content, width);
        bottomContent = (0, utils_mjs_1.breakLines)(bottomContent, width);
        // Manually insert an extra line if we're at the end of the line.
        // This prevent the cursor from appearing at the beginning of the
        // current line.
        if (rawPromptLine.length % width === 0) {
            content += '\n';
        }
        const fullContent = content + (bottomContent ? '\n' + bottomContent : '');
        this.rl.output.write(fullContent);
        /**
         * Re-adjust the cursor at the correct position.
         */
        // We need to consider parts of the prompt under the cursor as part of the bottom
        // content in order to correctly cleanup and re-render.
        const promptLineUpDiff = Math.floor(rawPromptLine.length / width) - cursorPos.rows;
        const bottomContentHeight = promptLineUpDiff + (bottomContent ? height(bottomContent) : 0);
        // Return cursor to the input position (on top of the bottomContent)
        util.up(this.rl, bottomContentHeight);
        // Move cursor at the start of the line, then return to the initial left offset.
        util.left(this.rl, (0, string_width_1.default)(lastLine(fullContent)));
        util.right(this.rl, cursorPos.cols);
        /**
         * Set up state for next re-rendering
         */
        this.extraLinesUnderPrompt = bottomContentHeight;
        this.height = height(fullContent);
        this.rl.output.mute();
    }
    clean() {
        this.rl.output.unmute();
        util.down(this.rl, this.extraLinesUnderPrompt);
        util.clearLine(this.rl, this.height);
        this.extraLinesUnderPrompt = 0;
        this.rl.output.mute();
    }
    clearContent() {
        this.rl.output.unmute();
        // Reset the cursor at the end of the previously displayed content
        util.down(this.rl, this.extraLinesUnderPrompt);
        this.rl.output.write('\n');
        this.rl.output.mute();
    }
    done() {
        this.rl.setPrompt('');
        this.rl.output.unmute();
        this.rl.output.write(ansi_escapes_1.default.cursorShow);
        this.rl.output.end();
        this.rl.close();
    }
}
exports.default = ScreenManager;
